content/blog/2023-12-11-btor2i.md

```diff
@@ -0,0 +1,78 @@
++++
+title = "An Interpreter for the btor2 Model Checking Language"
+[extra]
+bio = """
+  Sanjit, Omkar, and Justin are all MEng students in CS at Cornell.
+"""
+[[extra.authors]]
+name = "Sanjit Basker"
+link = "https://github.com/SanjitBasker"
+[[extra.authors]]
+name = "Omkar Bhalerao"
+link = "https://github.com/obhalerao" 
+[[extra.authors]]
+name = "Justin Ngai"
+link = "https://github.com/NgaiJustin"
++++
+
+[Link to project repository](https://github.com/NgaiJustin/btor2i)
+
+## Abstract
+In this project, we implemented an interpreter for the btor2 hardware description language. As there does not currently exist an interpreter for btor2, we felt that creating one would add to the overall language ecosystem. In addition, we implemented an optimization to the memory format and benchmarked it against our initial naive interpreter. We indeed observed that the optimization boosted performance on all benchmarks we surveyed.
+
+## What is btor2?
+btor2 is a model checking language designed for hardware; it provides this functionality at a word level, and also includes functionality for persistent state. Words of arbitrary length can be represented as bit vectors, and states that are present in the hardware (for instance, in registers or in local memory) can be represented as arrays. btor2 is an extension of btor, which is a similar model checking language for hardware which only supports bit vectors and one-dimensional bit vector arrays (in contrast, btor2 can have multidimensional arrays). btor2 as a language is useful for verifying generated low-level hardware code.
+
+## Motivation
+
+The creators of btor2 included three key tools: a parser, a random state simulator, and a witness checker, all written in C. Notably missing from this list is an interpreter for the language; though the random state simulator does exist, it does not test any of the many bit vector operations that btor2 supports, choosing to focus on arrays instead. In addition, though the witness checker does check whether a given state is a valid execution state for a given program, it does not do so in a constructive way—no actual interpretation is performed. We also feel that an interpreter for btor2 would be useful, since it would allow users to run arbitrary btor2 programs to ensure their correctness. Thus, we elected to write an interpreter for btor2 as this project.
```

Comment:
This paragraph could go one step farther to crystallize what is currently missing: a way to _specify the inputs_ to a btor2 program (to run it and observe the outputs). None of the other modes let you tell it the input you want to use to run the program. Witnesses sorta do, but you also have to specify the output, which isn't very useful. :smiley: